libazureinit/src/goalstate.rs

```diff
@@ -41,34 +45,63 @@ pub struct RoleInstance {
     instance_id: String,
 }
 
-pub async fn get_goalstate(client: &Client) -> Result<Goalstate, Error> {
-    let url = "http://168.63.129.16/machine/?comp=goalstate";
+const DEFAULT_GOALSTATE_URL: &str =
+    "http://168.63.129.16/machine/?comp=goalstate";
+
+pub async fn get_goalstate(
+    client: &Client,
+    retry_interval: Duration,
+    total_timeout: Duration,
+    url: Option<&str>,
+) -> Result<Goalstate, Error> {
+    let url = url.unwrap_or(DEFAULT_GOALSTATE_URL);
 
     let mut headers = HeaderMap::new();
     headers.insert("x-ms-agent-name", HeaderValue::from_static("azure-init"));
     headers.insert("x-ms-version", HeaderValue::from_static("2012-11-30"));
 
-    let request = client.get(url).headers(headers);
-    let response = request.send().await?;
+    let response = timeout(total_timeout, async {
+        loop {
+            if let Ok(response) = client
+                .get(url)
+                .headers(headers.clone())
+                .timeout(Duration::from_secs(30))
+                .send()
+                .await
+            {
+                let statuscode = response.status();
 
-    if response.status().is_success() {
-        let body = response.text().await?;
+                if statuscode.is_success() && statuscode == StatusCode::OK {
+                    return Ok(response);
+                }
 
-        let goalstate: Goalstate = from_str(&body)?;
-        Ok(goalstate)
-    } else {
-        Err(Error::HttpStatus {
-            endpoint: url.to_owned(),
-            status: response.status(),
-        })
-    }
+                if http::RETRY_CODES.contains(&statuscode) {
```

Comment:
what if there's a connection/read timeout error and there's no http error code? we want to retry those as well. Same would be applicable for the imds retry

Reply:
Now that the condition became different, the issue mentioned is automatically resolved.
That means, as we have `if !http::RETRY_CODES.contains`, here we return immediately for `HARDFAIL_CODES`. However, in case of other status code like `RETRY_CODES`, fall back to sleep and retry below.